src/readers/streams_reader.c

```diff
@@ -957,99 +958,253 @@ static bool StreamReader_IsStream(RedisModuleKey *kp){
     return RedisModule_KeyType(kp) == REDISMODULE_KEYTYPE_STREAM;
 }
 
-static void* StreamReader_ScanForStreams(void* pd){
-    RedisGears_LockHanlderRegister();
-    StreamReaderTriggerCtx* srctx = pd;
-    while (!Cluster_IsInitialized()) {
-    	/* Wait until cluster is initialized,
-    	 * Notice that there might be an adge case
-    	 * where cluster will be initialized here but
-    	 * will get uninitialized when we try to read
-    	 * the stream, for this we also have retry mechanism */
-    	sleep(1);
-    }
-    long long cursor = 0;
-    do{
-        // we do not use the lockhandler cause this thread is temporary
-        // and we do not want to allocate any unneeded extra data.
-        LockHandler_Acquire(staticCtx);
-        RedisModuleCallReply *reply = RedisModule_Call(staticCtx, "SCAN", "lcccc", cursor, "COUNT", "10000", "MATCH", srctx->args->streamPrefix);
-        LockHandler_Release(staticCtx);
-
-        bool ret = StreamReader_VerifyCallReply(staticCtx, reply, "Failed scanning keys on background", "warning");
-        RedisModule_Assert(ret);
-
-        RedisModule_Assert(RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ARRAY);
+typedef struct ScanCtx {
+    Gears_list *keys;
+    char *pattern;
+} ScanCtx;
 
-        RedisModule_Assert(RedisModule_CallReplyLength(reply) == 2);
-
-        RedisModuleCallReply *cursorReply = RedisModule_CallReplyArrayElement(reply, 0);
-
-        RedisModule_Assert(RedisModule_CallReplyType(cursorReply) == REDISMODULE_REPLY_STRING);
-
-        RedisModuleString *cursorStr = RedisModule_CreateStringFromCallReply(cursorReply);
-        RedisModule_StringToLongLong(cursorStr, &cursor);
-        RedisModule_FreeString(NULL, cursorStr);
+static void StreamReader_ScanFreeKey(void* pd){
+    if (!pd) return;
+    RedisModuleString *keyname = pd;
+    RedisModule_FreeString(NULL, keyname);
+}
 
-        RedisModuleCallReply *keysReply = RedisModule_CallReplyArrayElement(reply, 1);
-        RedisModule_Assert(RedisModule_CallReplyType(keysReply) == REDISMODULE_REPLY_ARRAY);
-        if(RedisModule_CallReplyLength(keysReply) < 1){
-            RedisModule_FreeCallReply(reply);
-            continue;
+static int stringmatchlen(const char *pattern, int patternLen,const char *string, int stringLen, int nocase, int *skipLongerMatches, int nesting)
+{
+    /* Protection against abusive patterns. */
+    if (nesting > 1000) return 0;
```

Comment:
Also, if the `pattern`, `string` are NULL or empty, and if `patternLen` or `stringLen` is empty.

Reply:
This function is a copy past from Redis, because RM_Scan do not support pattern filtering I had to move this logic here. It is well tested in Redis and any change to this function is consider a risk.